src/MCPServer/lib/server/queues.py

```diff
@@ -221,34 +197,33 @@ def stop(self):
         self.shutdown_event.set()
 
     def _package_completed_callback(self, package, future):
-        """Called by an Executor on completion the last job in a package.
+        """Mark the package as inactive and schedule a new package.
 
-        The result of the future will always be None.
+        It is assumed that a package is only complete when a terminal link is
+        hit. When we hit a terminal link the future should not contain the next
+        job in the chain. If the future is a new job we raise an exception.
         """
```

Comment:
I like what you're doing here but I thought the bits about the execution context were also useful (Called by an Executor...). Didn't you find that useful when you read the method for the first time?

> If the future is a new job we raise an exception.

You've provided some useful context but I think we also have to draw a line where we stop commenting on *what* the code is doing because the code is self-describing. This idea I'm trying to communicate is better described [here](https://google.github.io/eng-practices/review/reviewer/looking-for.html#comments).

Reply:
>  I thought the bits about the execution context were also useful

Originally I saw this as commenting about an implementation detail of the Python futures. However, looking at it again I see that this detail is actually coupled to how you can use the function. I.e. you cannot use the function to complete a package without using futures etc. So I agree that this adds some additional context that is relevant to the developer.

> This idea I'm trying to communicate is better described here.

I fully agree, missed the mark there. It is trivial to see in the code that we do this, and we already explained why - so that part was definitely obsolete.